prog/test/github_runner.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+require "octokit"
+require "yaml"
+
+class Prog::Test::GithubRunner < Prog::Test::Base
+  FAIL_CONCLUSIONS = ["action_required", "cancelled", "failure", "skipped", "stale", "timed_out"]
+  IN_PROGRESS_CONCLUSIONS = ["in_progress", "queued", "requested", "waiting", "pending", "neutral"]
+
+  def self.assemble(vm_host_id, test_cases)
+    github_service_project = Project.create(name: "Github Runner Service Project") { _1.id = Config.github_runner_service_project_id }
+    github_service_project.associate_with_project(github_service_project)
+
+    github_test_project = Project.create_with_id(name: "Github Runner Test Project")
+    github_test_project.associate_with_project(github_test_project)
+    GithubInstallation.create_with_id(
+      installation_id: Config.e2e_github_installation_id,
+      name: "TestUser",
+      type: "User",
+      project_id: github_test_project.id
+    )
+
+    Strand.create_with_id(
+      prog: "Test::GithubRunner",
+      label: "start",
+      stack: [{
+        "vm_host_id" => vm_host_id,
+        "test_cases" => test_cases,
+        "github_service_project_id" => github_service_project.id,
+        "github_test_project_id" => github_test_project.id
+      }]
+    )
+  end
+
+  label def start
+    hop_download_boot_images
+  end
+
+  label def download_boot_images
+    frame["test_cases"].each do |test_case|
+      bud Prog::DownloadBootImage, {"subject_id" => vm_host_id, "image_name" => tests[test_case]["image_name"]}
+    end
+
+    hop_wait_download_boot_images
+  end
+
+  label def wait_download_boot_images
+    reap
+    hop_trigger_test_runs if leaf?
+    donate
+  end
+
+  label def trigger_test_runs
+    test_runs.each do |test_run|
+      unless trigger_test_run(test_run["repo_name"], test_run["workflow_name"], test_run["branch_name"])
+        update_stack({"fail_message" => "Can not trigger workflow for #{test_run["repo_name"]}, #{test_run["workflow_name"]}, #{test_run["branch_name"]}"})
+        hop_clean_resources
+      end
+    end
+
+    # To make sure that test runs are triggered
+    sleep 30
```

Comment:
30 seconds may not be enough, especially during frequent GitHub incidents. To ensure safety, we can check the creation date of the latest run to confirm it was created after our trigger. If a new run isn't created, it will use the previous one and assume success. The project's creation time can serve as an example. You can either save the timestamp to the frame or use the one from the project which it's not perfect one. This check can also be performed under the `check_test_runs` label.